Philips.CodeAnalysis.MaintainabilityAnalyzers/Cardinality/AvoidVoidReturnAnalyzer.cs

```diff
@@ -0,0 +1,54 @@
+﻿// © 2023 Koninklijke Philips N.V. See License.md in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using LanguageExt;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Philips.CodeAnalysis.Common;
+using static LanguageExt.Prelude;
+using static Philips.CodeAnalysis.MaintainabilityAnalyzers.Cardinality.MethodPredicates;
+
+namespace Philips.CodeAnalysis.MaintainabilityAnalyzers.Cardinality
+{
+
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public class AvoidVoidReturnAnalyzer : DiagnosticAnalyzer
+	{
+
+		private const string Title = @"Method returns void";
+		private const string MessageFormat = @"Method {0} returns void";
+		private const string Description = @"Void returns imply a hidden side effect, since there is otherwise a singularly unique unit function.";
+		private const string Category = "Functional Programming";
+
+		public static readonly DiagnosticDescriptor Rule = new(Helper.ToDiagnosticId(DiagnosticId.AvoidVoidReturn), Title, MessageFormat, Category, DiagnosticSeverity.Error, isEnabledByDefault: false, description: Description);
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Rule); } }
+
+		private static readonly Action<SyntaxNodeAnalysisContext> ReportAnalysisAction =
```

Comment:
I don't understand this field. Why wouldn't Iter(cntx.ReportDiagnostic) just be appended to AnalyzeMethod? It almost looks like we're jumping through a complicated hoop just to avoid returning void. But if we append to AnalyzeMethod then we can override Analyze from the new SingleDiagnosticAnalyzer<T,U> class.  (And maybe from a FP perspective it's better to have Analyze return Optional<Diagnostic>, and then the base class can Report it?)